src/librustc_mir/borrow_check/path_utils.rs

```diff
@@ -61,6 +61,16 @@ pub(super) fn each_borrow_involving_path<'a, 'tcx, 'gcx: 'tcx, F, I, S> (
     for i in candidates {
         let borrowed = &borrow_set[i];
 
+        let access = if borrowed.kind == BorrowKind::Match {
+            // Prevent matching on *x and assigning to x in a guard.
+            // Match borrows are different to other borrows in that they exist
+            // to prevent the value of a Place being modified, rather than the
+            // value of a memory location.
+            ShallowOrDeep::Deep
+        } else {
+            access
```

Comment:
It is weird to me that we examine the borrow kind in order to change the access. I see why it sort of does the right thing, but something seems off to me. I wonder if what we "really want" here is instead to have multiple shared borrows, rather than adding a new kind of borrow -- that is, we originally thought that a single borrow of the "place being matched" would suffice, but it seems like that is not true, we want to borrow all prefixes of the place being matched perhaps? Or..something like that.

Reply:
@matthewjasper 

> But there still needs to be some way to not error when matching on a borrowed place.

I'm not entirely sure what you mean by that, actually, what's the example you have in mind?

Reply:
[This example from #53114](https://play.rust-lang.org/?gist=0e487c2893b89cb772ec2f2b7c5da876&version=nightly&mode=debug&edition=2015)